src/main/scala/components/memory-async.scala

```diff
@@ -0,0 +1,378 @@
+ *
+ * Output: response, the valid interface for the data outputted by memory if it was requested to read. 
+ *         the bits in response.bits should only be treated as valid data when response.valid is high.
+ */
+class AsyncMemIO extends Bundle {
+  val request  = Flipped(Decoupled (new Request))
+  val response = Valid (UInt (32.W))
+}
+
+/** 
+ * A generic ready/valid interface for MemAccess modules, whose IOs extend this.
+ * Input:  address, the address of a piece of data in memory. 
+ * Input:  valid, true when the address specified is valid
+ * Input:  response, the return route from memory to a memory access. This is primarily meant for connecting to 
+ *         an AsyncMemIO's response output, and should not be connected to anything else in any circumstance 
+ *         (or things will possibly break)
+ * 
+ * Output: ready, true when memory is either idling or outputting a value, and is ready for a new 
+ *         request. Note that this is different from access_out.ready - this ready is for use by the general
+ *         CPU (like to signal when to stall the CPU), while access_out.ready is used for signalling 
+ *         between the memory and accessor only
+ * Output: request, a DecoupledIO that delivers a request from a memory accessor to memory. This is primarily
+ *         meant for connecting to an AsynMemIO's request input, and should not be connected to anything else
+ */
+class MemAccessIO extends Bundle {
+  val address  = Input(UInt(32.W))
+  val valid    = Input(Bool())
+  val response = Flipped(Valid(UInt(32.W)))
+
+  val ready    = Output(Bool())
+  val request  = Decoupled(new Request)
+}
+
+/** 
+ * The *interface* of the IMemAccess module.
+ *
+ * Input:  address, the address of an instruction in memory 
+ * Input:  valid, true when the address specified is valid
+ *
+ * Output: instruction, the requested instruction
+ * Output: ready, true when memory is idling and ready for a request
+ */
+class IMemAccessIO extends MemAccessIO {
+  val instruction = Output(UInt(32.W))
+}
+
+/**
+ * The *interface* of the DMemAccess module.
+ *
+ * Input:  address, the address of a piece of data in memory. 
+ * Input:  writedata, valid interface for the data to write to the address
+ * Input:  valid, true when the address (and writedata during a write) specified is valid
+ * Input:  memread, true if we are reading from memory
+ * Input:  memwrite, true if we are writing to memory
+ * Input:  maskmode, mode to mask the result. 0 means byte, 1 means halfword, 2 means word
+ * Input:  sext, true if we should sign extend the result
+ *
+ * Output: readdata, the data read and sign extended
+ * Output: ready, true when memory is idling and ready for a request
+ */
+class DMemAccessIO extends MemAccessIO {
+  val writedata = Input(UInt(32.W))
+  val memread   = Input(Bool())
+  val memwrite  = Input(Bool())
+  val maskmode  = Input(UInt(2.W))
+  val sext      = Input(Bool())
+
+  val readdata  = Output(UInt(32.W))
+}
+
+/**
```

Comment:
The `IMemAccess` and `DMemAccess` could be moved to their own file (`sync-mem-ports`?)